plenum/server/replica.py

```diff
@@ -1688,6 +1706,87 @@ def can_pp_seq_no_be_in_view(self, view_no, pp_seq_no):
                                           compare_3PC_keys((view_no, pp_seq_no),
                                                            self.last_prepared_before_view_change) >= 0)
 
+    def request_pre_prepare_if_possible(self, three_pc_key) -> bool:
+        """
+        Check if has an acceptable PRE_PREPARE already stashed, if not then
+        check count of PREPAREs, make sure >f consistent PREPAREs are found,
+        store the acceptable PREPARE state (digest, roots) for verification of
+        the received PRE-PREPARE
+        """
+        if len(self.preparesWaitingForPrePrepare[three_pc_key]) < self.quorums.prepare.value:
+            logger.debug('{} not requesting a PRE-PREPARE because does not have'
+                         ' sufficient PREPAREs for {}'.format(self, three_pc_key))
+            return False
+
+        if three_pc_key in self.requested_pre_prepares:
+            logger.debug('{} not requesting a PRE-PREPARE since already '
+                         'requested for {}'.format(self, three_pc_key))
+            return False
+
+        if three_pc_key in self.prePreparesPendingPrevPP:
+            logger.debug('{} not requesting a PRE-PREPARE since already found '
+                         'stashed for {}'.format(self, three_pc_key))
+            return False
+
+        # Choose a better data structure for `prePreparesPendingFinReqs`
+        pre_prepares = [pp for pp, _, _ in self.prePreparesPendingFinReqs
+                        if (pp.viewNo, pp.ppSeqNo) == three_pc_key]
+        digest, state_root, txn_root, prepare_senders = \
+            self.get_acceptable_stashed_prepare_state(three_pc_key)
+        if pre_prepares:
+            if [pp for pp in pre_prepares if
+                (pp.digest, pp.stateRootHash, pp.txnRootHash) == (digest, state_root, txn_root)]:
+                logger.debug('{} not requesting a PRE-PREPARE since already '
+                             'found stashed for {}'.format(self, three_pc_key))
+                return False
+
+        # TODO: Using a timer to retry would be a better thing to do
+        logger.debug('{} requesting PRE-PREPARE({}) from {}'.
+                     format(self, three_pc_key, prepare_senders))
+        self.node.request_msg(PREPREPARE, {f.INST_ID.nm: self.instId,
```

Comment:
I don't agree with this. We can not implement good contracts with a request any message approach.
We can not use validation properly. It's absolutely not clear what parameters should be passed to each request.
Why don't use separate messages for each of the Request Types? Explicit is better than implicit. We don't have hundreds of request, so there will be not so many explicit request messages